<div style="text-align: center; background-color: #0A6EBD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Final Project - Introduction To Data Science @ FIT-HCMUS, VNU-HCM 📌
</div>

<div style="text-align: center; background-color: #77F0A1; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 02 - Exploring Data 🔎
</div>

# Import

In [1]:
import numpy as np
import pandas as pd

# Exploring data

Read raw data from file

In [2]:
raw_df = pd.read_csv("../Data/raw/popular_git_repositories.csv")
raw_df.head()

Name   
0                 freeCodeCamp  \
1       free-programming-books   
2                      awesome   
3                      996.ICU   
4  coding-interview-university   

                                         Description   
0  freeCodeCamp.org's open-source codebase and cu...  \
1         :books: Freely available programming books   
2  😎 Awesome lists about all kinds of interesting...   
3  Repo for counting stars and contributing. Pres...   
4  A complete computer science study plan to beco...   

                                                 URL            Created At   
0       https://github.com/freeCodeCamp/freeCodeCamp  2014-12-24T17:49:19Z  \
1  https://github.com/EbookFoundation/free-progra...  2013-10-11T06:50:37Z   
2            https://github.com/sindresorhus/awesome  2014-07-11T13:42:37Z   
3                  https://github.com/996icu/996.ICU  2019-03-26T07:31:14Z   
4  https://github.com/jwasham/coding-interview-un...  2016-06-06T02:34:12Z   

             Updated At                                           Homepage   
0  2023-09-21T11:32:33Z                http://contribute.freecodecamp.org/  \
1  2023-09-21T11:09:25Z  https://ebookfoundation.github.io/free-program...   
2  2023-09-21T11:18:22Z                                                NaN   
3  2023-09-21T08:09:01Z                                    https://996.icu   
4  2023-09-21T10:54:48Z                                                NaN   

     Size   Stars  Forks  Issues  ...  Has Issues Has Projects Has Downloads   
0  387451  374074  33599     248  ...        True         True          True  \
1   17087  298393  57194      46  ...        True        False          True   
2    1441  269997  26485      61  ...        True        False          True   
3  187799  267901  21497   16712  ...       False        False          True   
4   20998  265161  69434      56  ...        True        False          True   

  Has Wiki  Has Pages  Has Discussions  Is Fork  Is Archived  Is Template   
0    False       True            False    False        False        False  \
1    False       True            False    False        False        False   
2    False       True            False    False        False        False   
3    False      False            False    False         True        False   
4    False      False            False    False        False        False   

   Default Branch  
0            main  
1            main  
2            main  
3          master  
4            main  

[5 rows x 24 columns]

## How many rows and how many columns does the raw data have?

In [3]:
shape = raw_df.shape
print(f"Current shape: {shape}")

Current shape: (215029, 24)


## What is the meaning of each row?

**Answer:** 
- Based on the observations of the dataset, we suppose that each row corresponds to a distinct information about various aspects of a specific GitHub repository.

## Does the raw data have duplicate rows?

In [4]:
# retrieve the index
index = raw_df.index
# create a Pandas Series indicating whether each index is duplicated or not
deDupSeries = index.duplicated(keep='first')
# calculate the number of duplicated rows
num_duplicated_rows = deDupSeries.sum()

In [5]:
if num_duplicated_rows == 0:
    print(f"Raw data have no duplicated line !")
else:
    if num_duplicated_rows > 1:
        ext = "lines"
    else:
        ext = "line"
    print(f"Raw data have {num_duplicated_rows} duplicated " + ext)

Raw data have no duplicated line !


## What does each column mean?


<table style="width: 100%; border-collapse: collapse;">
  <thead>
    <tr>
      <th style="min-width: 10px; width: 30px; background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">No</th>
      <th style="background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">Columns</th>
      <th style="background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">Meaning</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="font-size: 14px; text-align: left;">1</td>
      <td style="font-size: 14px; text-align: left;">Name</td>
      <td style="font-size: 14px; text-align: left;">The name of the GitHub repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">2</td>
      <td style="font-size: 14px; text-align: left;">Description</td>
      <td style="font-size: 14px; text-align: left;">A brief description of the repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">3</td>
      <td style="font-size: 14px; text-align: left;">URL</td>
      <td style="font-size: 14px; text-align: left;">The URL to access the repository on GitHub.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">4</td>
      <td style="font-size: 14px; text-align: left;">Created At</td>
      <td style="font-size: 14px; text-align: left;">The time indicating when the repository was created.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">5</td>
      <td style="font-size: 14px; text-align: left;">Updated At</td>
      <td style="font-size: 14px; text-align: left;">The time indicating when the repository was last updated.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">6</td>
      <td style="font-size: 14px; text-align: left;">Homepage</td>
      <td style="font-size: 14px; text-align: left;">The URL of the repository's homepage or website.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">7</td>
      <td style="font-size: 14px; text-align: left;">Size</td>
      <td style="font-size: 14px; text-align: left;">The size of the repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">8</td>
      <td style="font-size: 14px; text-align: left;">Stars</td>
      <td style="font-size: 14px; text-align: left;">The number of stars the repository has received.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">9</td>
      <td style="font-size: 14px; text-align: left;">Forks</td>
      <td style="font-size: 14px; text-align: left;">The number of times the repository has been forked (copied).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">10</td>
      <td style="font-size: 14px; text-align: left;">Issues</td>
      <td style="font-size: 14px; text-align: left;">The number of open issues in the repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">11</td>
      <td style="font-size: 14px; text-align: left;">Watchers</td>
      <td style="font-size: 14px; text-align: left;">The number of users who are watching the repository for updates.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">12</td>
      <td style="font-size: 14px; text-align: left;">Language</td>
      <td style="font-size: 14px; text-align: left;">The primary programming language used in the repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">13</td>
      <td style="font-size: 14px; text-align: left;">License</td>
      <td style="font-size: 14px; text-align: left;">The type of license under which the repository's code is distributed.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">14</td>
      <td style="font-size: 14px; text-align: left;">Topics</td>
      <td style="font-size: 14px; text-align: left;">A list of topics or tags associated with the repository.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">15</td>
      <td style="font-size: 14px; text-align: left;">Has Issues</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has an issue tracker (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">16</td>
      <td style="font-size: 14px; text-align: left;">Has Projects</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has GitHub Projects enabled (True or False).</td> 
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">17</td>
      <td style="font-size: 14px; text-align: left;">Has Downloads</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has downloads enabled (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">18</td>
      <td style="font-size: 14px; text-align: left;">Has Wiki</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has a wiki (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">19</td>
      <td style="font-size: 14px; text-align: left;">Has Pages</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has GitHub Pages enabled (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">20</td>
      <td style="font-size: 14px; text-align: left;">Has Discussions</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository has GitHub Discussions enabled (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">21</td>
      <td style="font-size: 14px; text-align: left;">Is Fork</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository is a fork of another repository (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">22</td>
      <td style="font-size: 14px; text-align: left;">Is Archived</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository is archived (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">23</td>
      <td style="font-size: 14px; text-align: left;">Is Template</td>
      <td style="font-size: 14px; text-align: left;">Indicates whether the repository is a template (True or False).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">24</td>
      <td style="font-size: 14px; text-align: left;">Default Branch</td>
      <td style="font-size: 14px; text-align: left;">The default branch of the repository.</td>
    </tr>
  </tbody>
</table>

## What data type does each column currently have? Are there any columns having inappropriate data types?

Calculating the dtype of each column in raw_df

In [6]:
raw_df.dtypes

Name               object
Description        object
URL                object
Created At         object
Updated At         object
Homepage           object
Size                int64
Stars               int64
Forks               int64
Issues              int64
Watchers            int64
Language           object
License            object
Topics             object
Has Issues           bool
Has Projects         bool
Has Downloads        bool
Has Wiki             bool
Has Pages            bool
Has Discussions      bool
Is Fork              bool
Is Archived          bool
Is Template          bool
Default Branch     object
dtype: object

We notice that the columns `Created At` and `Updated At` are currently of object type. Given that they represent time periods, it is advisable to convert them to datetime type.

In [7]:
# convert to datetime
raw_df["Created At"] = pd.to_datetime(raw_df["Created At"])
raw_df["Updated At"] = pd.to_datetime(raw_df["Updated At"])

In [8]:
# TEST
raw_df.dtypes

Name                            object
Description                     object
URL                             object
Created At         datetime64[ns, UTC]
Updated At         datetime64[ns, UTC]
Homepage                        object
Size                             int64
Stars                            int64
Forks                            int64
Issues                           int64
Watchers                         int64
Language                        object
License                         object
Topics                          object
Has Issues                        bool
Has Projects                      bool
Has Downloads                     bool
Has Wiki                          bool
Has Pages                         bool
Has Discussions                   bool
Is Fork                           bool
Is Archived                       bool
Is Template                       bool
Default Branch                  object
dtype: object

## Are there any duplicated values?

In [9]:
print("Number of duplicated values:", raw_df.duplicated().sum())

Number of duplicated values: 0


## With each numerical column, how are values distributed?
- What is the percentage of missing values?
- Min? max? Are they abnormal?

### Percentage of missing values

In [10]:
missing_vals = raw_df.select_dtypes(include='number').isna().sum()
missing_percentage = missing_vals / len(raw_df)
missing_percentage

Size        0.0
Stars       0.0
Forks       0.0
Issues      0.0
Watchers    0.0
dtype: float64

### Min? Max? Are they abnormal?

In [11]:
raw_df.describe()

Size          Stars          Forks         Issues   
count  2.150290e+05  215029.000000  215029.000000  215029.000000  \
mean   5.428270e+04    1115.085142     234.207637      37.925178   
std    7.023978e+05    3992.372050    1242.967451     196.508610   
min    0.000000e+00     167.000000       0.000000       0.000000   
25%    3.780000e+02     237.000000      39.000000       3.000000   
50%    2.389000e+03     377.000000      79.000000      10.000000   
75%    1.528200e+04     797.000000     174.000000      28.000000   
max    1.050786e+08  374074.000000  243339.000000   26543.000000   

            Watchers  
count  215029.000000  
mean     1115.085142  
std      3992.372050  
min       167.000000  
25%       237.000000  
50%       377.000000  
75%       797.000000  
max    374074.000000

- For the 4 features **Stars, Forks, Issues and Watchers**, the *max* value seems to be very big compared to the rest of its column's data, 100 times larger than the 75% quantile.

- But, this is **understandable** because it means there are some repositories that are *hugely more popular* than most other repositories.

## With each categorical column, how are values distributed?
- What is the percentage of missing values?
- How many different values? Show a few
- Are they abnormal?

### What is the percentage of missing values?

In [12]:
missing_vals = raw_df.select_dtypes(exclude='number').isna().sum()
missing_percentage = missing_vals / len(raw_df)
missing_percentage

Name               0.000009
Description        0.037353
URL                0.000000
Created At         0.000000
Updated At         0.000000
Homepage           0.635445
Language           0.074762
License            0.246660
Topics             0.000000
Has Issues         0.000000
Has Projects       0.000000
Has Downloads      0.000000
Has Wiki           0.000000
Has Pages          0.000000
Has Discussions    0.000000
Is Fork            0.000000
Is Archived        0.000000
Is Template        0.000000
Default Branch     0.000000
dtype: float64

- Features with missing values are **Description, Homepage, Language and License.**

- Notably, **Homepage** has a missing percentage of **0.635445** (over half of the data).

### How many different values? Show a few

In [13]:
categorical_vals = raw_df.select_dtypes(exclude='number')

unique_vals = categorical_vals.apply(pd.Series.unique, axis=0)
unique_counts = unique_vals.apply(len)

Number of unique values for each column

In [14]:
print(unique_counts)

Name               196821
Description        206110
URL                215029
Created At         214922
Updated At         193011
Homepage            74198
Language              370
License                46
Topics             110123
Has Issues              2
Has Projects            2
Has Downloads           2
Has Wiki                2
Has Pages               2
Has Discussions         2
Is Fork                 1
Is Archived             2
Is Template             2
Default Branch       2326
dtype: int64


Some values for each column

In [15]:
print(unique_vals)

Name               [freeCodeCamp, free-programming-books, awesome...
Description        [freeCodeCamp.org's open-source codebase and c...
URL                [https://github.com/freeCodeCamp/freeCodeCamp,...
Created At         [2014-12-24 17:49:19+00:00, 2013-10-11 06:50:3...
Updated At         [2023-09-21 11:32:33+00:00, 2023-09-21 11:09:2...
Homepage           [http://contribute.freecodecamp.org/, https://...
Language           [TypeScript, nan, Python, JavaScript, C++, She...
License            [BSD-3-Clause, CC-BY-4.0, CC0-1.0, NOASSERTION...
Topics             [['careers', 'certification', 'community', 'cu...
Has Issues                                             [True, False]
Has Projects                                           [True, False]
Has Downloads                                          [True, False]
Has Wiki                                               [False, True]
Has Pages                                              [True, False]
Has Discussions                   

### Are they abnormal?

- **Is Fork** has only 1 value: False. We can remove this feature.

In [16]:
raw_df = raw_df.drop('Is Fork', axis=1)

## Save the processed data

In [17]:
# Save processed data to disk
raw_df.to_csv("../data/processed/" + "data_processed.csv", index=False)